861 segresnet

yiheng-wang-nv opened the pull request:
Implements #861 .

### Description
Under default settings (except utilizing batch norm), the model gets mean dice 0.9377 after 342 epochs in Spleen Dataset.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented on monai/networks/blocks/segresnet_block.py:
```diff
@@ -0,0 +1,101 @@
+
+    conv_type: Type[Union[nn.Conv1d, nn.Conv2d, nn.Conv3d]] = Conv[Conv.CONV, spatial_dims]
+
+    conv = conv_type(
```
could line 38-42 be implemented with `monai.networks.blocks.Convolution`?

yiheng-wang-nv replied:
Thanks! I changed it.

wyli commented on monai/networks/blocks/segresnet_block.py:
```diff
@@ -0,0 +1,101 @@
+
+class ResBlock(nn.Module):
+    """
+    ResBlock emploies skip connection and two convolution blocks and is used
```
typo: emploies  -> employs

yiheng-wang-nv replied:
thanks!

Nic-Ma commented:
Hi @wyli , 

We implemented this version based on the latest SegResNet implementation Andriy just sent us.
I can send you his original implementation if you like, he suggested to remove the VAE branch.

Thanks.

wyli commented:
> Hi @wyli ,
> 
> We implemented this version based on the latest SegResNet implementation Andriy just sent us.
> I can send you his original implementation if you like, he suggested to remove the VAE branch.
> 
> Thanks.

Ok, I thought the goal here is to try our best to reimplement what the paper describes.
Plus implementing the generic VAE logic could be a very nice addition to MONAI
 cc @myron

Nic-Ma commented:
Hi @wyli ,

@myron said the VAE part is not necessary anymore, need @myron to help provide more information here.
Thanks.

yiheng-wang-nv commented:
> Hi @wyli ,
> 
> @myron said the VAE part is not necessary anymore, need @myron to help provide more information here.
> Thanks.

Hi @Nic-Ma @wyli , I've added VAE parts and updated the code.

yiheng-wang-nv commented:
I see, let me refactor it

yiheng-wang-nv commented:
> thanks, looks good to me, please add `SegResnetVAE` to `docs/source/networks.rst`, then I think this is ready

I see, I add it.

wyli commented:
/black
